cpp/src/arrow/compute/kernels/comparator.h

```diff
@@ -0,0 +1,117 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class Array;
+class DataType;
+class Scalar;
+class Status;
+
+namespace compute {
+
+struct Datum;
+class FilterFunction;
+class FunctionContext;
+
+enum CompareOperator {
```

Comment:
Should these be scoped in a struct, as with `arrow::Type`?

Reply:
We'll see if needed, they could be used elsewhere in the codebase (logical operators).

Reply:
I just meant that our pattern for potentially public enums like this appears to be:
```c++
struct CompareOperator {
  enum type {
    // ...
```